File: backend/internal/handlers/resume.go
package handlers

import (
	"encoding/json"
	"net/http"
	"strconv"
	"strings"

	"resume-ai-backend/internal/models"
	"resume-ai-backend/internal/services"

	"github.com/gin-gonic/gin"
)

// toJSONString 将数据转换为 JSON 字符串
func toJSONString(data interface{}) string {
	if data == nil {
		return "{}"
	}

	switch v := data.(type) {
	case map[string]interface{}:
		if len(v) == 0 {
			return "{}"
		}
	case []map[string]interface{}:
		if len(v) == 0 {
			return "[]"
		}
	case []interface{}:
		if len(v) == 0 {
			return "[]"
		}
	}

	bytes, _ := json.Marshal(data)
	return string(bytes)
}

func GetResumes(c *gin.Context) {
	userID, ok := GetCurrentUserID(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	resumes, err := models.GetResumesByUserID(userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code": 0,
		"data": gin.H{
			"list":      resumes,
			"total":     len(resumes),
			"page":      1,
			"page_size": 10,
		},
	})
}

func GetResume(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid id"})
		return
	}

	resume, err := models.GetResumeByID(uint(id))
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "resume not found"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code": 0,
		"data": resume,
	})
}

func CreateResume(c *gin.Context) {
	var req struct {
		Title      string `json:"title"`
		ResumeType string `json:"resume_type"`
		UserID     uint   `json:"user_id"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if req.Title == "" {
		req.Title = "我的简历"
	}
	if req.ResumeType == "" {
		req.ResumeType = "full"
	}

	resume := &models.Resume{
		Title:          req.Title,
		ResumeType:     req.ResumeType,
		UserID:         req.UserID,
		Status:         "draft",
		PersonalInfo:   "{}",    // JSON 对象
		Education:      "[]",    // JSON 数组
		WorkExperience: "[]",
		Projects:       "[]",
		Skills:         "[]",
		Awards:         "[]",
		Languages:      "[]",
	}

	if err := models.CreateResume(resume); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code": 0,
		"data": resume,
	})
}

func UpdateResume(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid id"})
		return
	}

	resume, err := models.GetResumeByID(uint(id))
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "resume not found"})
		return
	}

	var req struct {
		Title          string `json:"title"`
		ThemeID        *int   `json:"theme_id"`
		ResumeType     string `json:"resume_type"`
		Status         string `json:"status"`
		IsDefault      *bool  `json:"is_default"`
		PersonalInfo   string `json:"personal_info"`
		Education      string `json:"education"`
		WorkExperience string `json:"work_experience"`
		Projects       string `json:"projects"`
		Skills         string `json:"skills"`
		Awards         string `json:"awards"`
		Languages      string `json:"languages"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if req.Title != "" {
		resume.Title = req.Title
	}
	if req.ThemeID != nil {
		resume.ThemeID = *req.ThemeID
	}
	if req.ResumeType != "" {
		resume.ResumeType = req.ResumeType
	}
	if req.Status != "" {
		resume.Status = req.Status
	}
	if req.PersonalInfo != "" {
		resume.PersonalInfo = req.PersonalInfo
	}
	if req.Education != "" {
		resume.Education = req.Education
	}
	if req.WorkExperience != "" {
		resume.WorkExperience = req.WorkExperience
	}
	if req.Projects != "" {
		resume.Projects = req.Projects
	}
	if req.Skills != "" {
		resume.Skills = req.Skills
	}
	if req.Awards != "" {
		resume.Awards = req.Awards
	}
	if req.Languages != "" {
		resume.Languages = req.Languages
	}

	if req.IsDefault != nil && *req.IsDefault {
		models.SetDefaultResume(resume.UserID, resume.ID)
	}

	if err := models.UpdateResume(resume); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code": 0,
		"data": resume,
	})
}

func DeleteResume(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid id"})
		return
	}

	if err := models.DeleteResume(uint(id)); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code":    0,
		"message": "deleted successfully",
	})
}

func OptimizeResume(c *gin.Context) {
	var req struct {
		Content string `json:"content" binding:"required"`
		Type    string `json:"type"`
		Target  string `json:"target"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	result, err := services.OptimizeWithAI(req.Content, req.Type, req.Target)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code": 0,
		"data": gin.H{
			"result": result,
		},
	})
}

// ImportResume 从 PDF 导入简历
func ImportResume(c *gin.Context) {
	userID, ok := GetCurrentUserID(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	// 获取上传的文件
	file, header, err := c.Request.FormFile("file")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "请上传文件"})
		return
	}
	defer file.Close()

	// 检查文件后缀
	if !strings.HasSuffix(strings.ToLower(header.Filename), ".pdf") {
		c.JSON(http.StatusBadRequest, gin.H{"error": "请上传 PDF 格式的文件"})
		return
	}

	// 检查文件大小 (最大 10MB)
	if header.Size > 10*1024*1024 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "文件大小不能超过 10MB"})
		return
	}

	// 解析 PDF
	parsedResume, err := services.ParsePDF(file)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "PDF 解析失败: " + err.Error()})
		return
	}

	// 转换为 JSON 字符串
	resumeData := parsedResume.ConvertToResumeData()

	// 创建新简历
	resume := &models.Resume{
		Title:          "导入的简历",
		ResumeType:     "full",
		UserID:         userID,
		Status:         "draft",
		PersonalInfo:   toJSONString(resumeData["personal_info"]),
		Education:      toJSONString(resumeData["education"]),
		WorkExperience: toJSONString(resumeData["work_experience"]),
		Projects:       toJSONString(resumeData["projects"]),
		Skills:         toJSONString(resumeData["skills"]),
		Awards:         "[]",
		Languages:      "[]",
	}

	// 使用解析出的姓名作为标题
	if personalInfo, ok := resumeData["personal_info"].(map[string]interface{}); ok {
		if name, ok := personalInfo["name"].(string); ok && name != "" {
			resume.Title = name + "的简历"
		}
	}

	if err := models.CreateResume(resume); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "保存简历失败: " + err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code":    0,
		"message": "导入成功",
		"data": gin.H{
			"resume":   resume,
			"raw_text": parsedResume.RawText,
			"parsed":   resumeData,
		},
	})
}

File: backend/schemas/resume.py
"""
简历相关 Pydantic 模型 - 与 Go 版本入参出参一致
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Any
from datetime import datetime


# ============ 请求模型 ============

class ResumeCreateRequest(BaseModel):
    """创建简历请求"""
    title: Optional[str] = "我的简历"
    theme_id: Optional[int] = 0
    resume_type: Optional[str] = "full"
    user_id: int  # 必填，与 Go 版本一致
    personal_info: Optional[str] = "{}"  # JSON 字符串
    education: Optional[str] = "[]"
    work_experience: Optional[str] = "[]"
    projects: Optional[str] = "[]"
    skills: Optional[str] = "[]"
    awards: Optional[str] = "[]"
    languages: Optional[str] = "[]"


class ResumeUpdateRequest(BaseModel):
    """更新简历请求"""
    title: Optional[str] = None
    theme_id: Optional[int] = None
    resume_type: Optional[str] = None
    status: Optional[str] = None
    is_default: Optional[bool] = None
    personal_info: Optional[str] = None
    education: Optional[str] = None
    work_experience: Optional[str] = None
    projects: Optional[str] = None
    skills: Optional[str] = None
    awards: Optional[str] = None
    languages: Optional[str] = None


# ============ 响应模型 ============

class ResumeBase(BaseModel):
    """简历基础字段"""
    id: int
    user_id: int
    title: str
    theme_id: int = 0
    resume_type: str = "full"
    is_default: bool = False
    status: str = "draft"
    personal_info: Any = {}
    education: Any = []
    work_experience: Any = []
    projects: Any = []
    skills: Any = []
    awards: Any = []
    languages: Any = []
    share_token: Optional[str] = None
    share_expires_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ResumeResponse(BaseModel):
    """简历响应"""
    code: int = 0
    data: ResumeBase


class ResumeListItem(BaseModel):
    """简历列表项（包含所有字段，与 Go 版本一致）"""
    id: int
    user_id: int
    title: str
    theme_id: int = 0
    resume_type: str = "full"
    is_default: bool = False
    status: str = "draft"
    # JSON 字符串字段
    personal_info: str = "{}"
    education: str = "[]"
    work_experience: str = "[]"
    projects: str = "[]"
    skills: str = "[]"
    awards: str = "[]"
    languages: str = "[]"
    share_token: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ResumeListData(BaseModel):
    """简历列表数据"""
    list: List[ResumeListItem]
    total: int
    page: int
    page_size: int


class ResumeListResponse(BaseModel):
    """简历列表响应"""
    code: int = 0
    data: ResumeListData


class ParsedResumeData(BaseModel):
    """解析后的简历数据"""
    personal_info: dict = {}
    education: List[dict] = []
    work_experience: List[dict] = []
    projects: List[dict] = []
    skills: List[dict] = []
    awards: List[dict] = []
    languages: List[dict] = []


class ImportData(BaseModel):
    """导入返回数据"""
    resume: ResumeBase
    raw_text: str
    parsed: ParsedResumeData


class ResumeImportResponse(BaseModel):
    """简历导入响应"""
    code: int = 0
    message: str = "导入成功"
    data: ImportData


class ShareResponse(BaseModel):
    """分享响应"""
    code: int = 0
    data: dict  # {share_token, share_url}


class DeleteResponse(BaseModel):
    """删除响应"""
    code: int = 0
    message: str = "deleted successfully"
